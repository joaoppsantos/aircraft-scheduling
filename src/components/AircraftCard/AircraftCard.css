.aircraft-card {
    height: 10vh;
    width: 100%;
    font-size: 24px;
    margin-top: 14px;
    background-color: lightblue;

    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
}
