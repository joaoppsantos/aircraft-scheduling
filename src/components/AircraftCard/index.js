export { default as AircraftCard } from './AircraftCard';
