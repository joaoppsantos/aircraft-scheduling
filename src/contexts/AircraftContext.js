import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
