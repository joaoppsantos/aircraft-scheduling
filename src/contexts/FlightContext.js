import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});