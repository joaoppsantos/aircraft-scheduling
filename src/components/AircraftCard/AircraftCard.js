import './AircraftCard.css';
import { useContext, useEffect } from 'react';
import { AircraftContext } from '../../contexts/AircraftContext';
import { FlightContext } from '../../contexts/FlightContext';

const AircraftCard = aircraft => {
	const {
		selectAircraft,
		currentPercentage,
		setPercentage
	} = useContext(AircraftContext);

	const {
		activeFlights
	} = useContext(FlightContext);

	const handleAircraftClick = () => {
		selectAircraft(aircraft);
	}

	let flightTime = 0, secondsInADay = 86400, sumPercentage = 0;

	useEffect(() => {
		sumPercentage = 0;

		activeFlights.forEach(flight => {
			flightTime = flight.arrivaltime - flight.departuretime;

			sumPercentage += (flightTime/secondsInADay) * 100;
		})

		setPercentage(Math.round(sumPercentage));
	}, [activeFlights])
	
	return (
		<div className="aircraft-card" onClick={handleAircraftClick}> 
			<span>{aircraft?.ident}</span>
			<span>{currentPercentage}%</span>
		</div>
	);
}

export default AircraftCard;
