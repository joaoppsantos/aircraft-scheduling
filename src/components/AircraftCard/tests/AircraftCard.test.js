import * as React from 'react';
import { render } from '@testing-library/react';
import AircraftCard from '../AircraftCard';
import { AircraftContext } from '../../../contexts/AircraftContext';
import { FlightContext } from '../../../contexts/FlightContext';

const flightMock = {
    "id": "AS1001",
    "departuretime": 21600,
    "arrivaltime": 26100,
    "readable_departure": "06:00",
    "readable_arrival": "07:15",
    "origin": "LFSB",
    "destination": "LFMN"
};
const mockAircraftProps = {
    ident: "GABCD",
    economySeats: 186,
    base: "EGKK",
    type: "A320",
};

describe('<AircraftCard />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            activeFlights: [
                flightMock,
            ],
        };
        const mockAircraftContext = {
            selectAircraft: jest.fn(),
            currentPercentage: "50",
            setPercentage: jest.fn()
        }

        const { container } = render(
            <AircraftContext.Provider value={mockAircraftContext}>
                <FlightContext.Provider value={mockFlightContext}>
                    <AircraftCard {...mockAircraftProps} />
                </FlightContext.Provider>
            </AircraftContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
