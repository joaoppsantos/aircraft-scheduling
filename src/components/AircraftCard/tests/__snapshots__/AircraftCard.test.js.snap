// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AircraftCard /> should render correctly 1`] = `
<div>
  <div
    class="aircraft-card"
  >
    <span>
      GABCD
    </span>
    <span>
      50
      %
    </span>
  </div>
</div>
`;
